src/index.tsx

```diff
@@ -325,23 +325,25 @@ export class App extends React.Component<any, AppState> {
       }
       this.setState({ elementType: shape });
     } else if (event[KEYS.META] && event.code === "KeyZ") {
+      event.preventDefault();
+
       if (event.shiftKey) {
         // Redo action
         const data = history.redoOnce();
         if (data !== null) {
-          this.setState(data.appState);
           elements = data.elements;
+          this.forceUpdate();
+          this.setState(data.appState);
         }
       } else {
         // undo action
         const data = history.undoOnce();
         if (data !== null) {
-          this.setState(data.appState);
           elements = data.elements;
+          this.forceUpdate();
+          this.setState(data.appState);
         }
```

Comment:
Does this re-render twice?

Reply:
As I wrote below, forcing refresh after `this.setState` would duplicate history entries.

Reply:
It still happens with state properties though.

Reply:
Okay... Either way I'd drop the `forceUpdate` for two reasons:

1. it's not needed in classical React (but yes, we use it elsewhere, too, for no reason)
2. more importantly, even if we *did* need to ensure re-render, I don't see any reason why forcing re-render *before* we change state would help in any way?

Reply:
Okay, there actually is a reason why re-rendering before state change does something: we modify history entries in `componentDidUpdate`.

Mhm, then I need to think what happens. Gimme a moment.

Reply:
Right right, I'm a bit sleepy sorry. Indeed we don't need to `forceUpdate` in this case even though we are updating `elements` which is outside of React's scope because we call `this.setState` 🤦‍♂ 

Looks like it's working perfectly now 👍

Reply:
> Okay, there actually is a reason why re-rendering before state change does something: we modify history entries in componentDidUpdate.

Yes, the app rerenders many times without a reason if you look at the entries we try to push into the history. Nothing happens history wise because we block pushing an entry equal to the previous one, but it is definitely not performant 😄

Reply:
Mhm.. not entirely sure what's going on, but either way it seems that the `forceUpdate` is a no-op in terms of history because it bails due to trying to push the same entry (Why, I'm not sure)

I'll need to review how the history works to ensure there are no other hidden bugs etc., but for now we can merge this. Removing the `forceUpdate` was a good idea after all because it not only did nothing in this case, it also was adding complexity that would end up taking someone else's brain power in the future.

Reply:
I rewrote the history in another PR recently, and I'm pretty convinced there are no bugs history. It's just about the state management in React and our use of `forceUpdate`.

Reply:
Yea, I'm tired too (we're in the same time zone). Good job. :rocket: